cluster.go

```diff
@@ -0,0 +1,316 @@
+		// and request again.
+		if _, ok := err.(*net.OpError); ok || err == io.EOF {
+			if addr = c.findNextAddr(tried); addr == "" {
+				return
+			}
+			cmd.reset()
+			continue
+		}
+
+		// Check the error message, return if unexpected
+		parts := strings.SplitN(err.Error(), " ", 3)
+		if len(parts) != 3 {
+			return
+		}
+
+		// Handle MOVE and ASK redirections, return on any other error
+		switch parts[0] {
+		case "MOVED":
+			c.forceReload()
+			addr = parts[2]
+		case "ASK":
+			ask = true
+			addr = parts[2]
+		default:
+			return
+		}
+		cmd.reset()
+	}
+}
+
+// Closes all connections and reloads slot cache, if due
+func (c *ClusterClient) reloadIfDue() (err error) {
+	if !atomic.CompareAndSwapUint32(&c._reload, 1, 0) {
+		return
+	}
+
+	var infos []ClusterSlotInfo
+
+	c.cachemx.Lock()
+	defer c.cachemx.Unlock()
+
+	// Try known addresses in random order (map interation order is random in Go)
+	// http://redis.io/topics/cluster-spec#clients-first-connection-and-handling-of-redirections
+	// https://github.com/antirez/redis-rb-cluster/blob/fd931ed/cluster.rb#L157
+	for addr := range c.addrs {
+		c.reset()
+
+		infos, err = c.fetchClusterSlots(addr)
+		if err == nil {
+			c.update(infos)
+			break
+		}
+	}
+	return
+}
+
+// Closes all connections and flushes slots cache
+func (c *ClusterClient) reset() error {
+	err := c.conns.Clear()
+	c.slots = make([][]string, hashSlots)
+	return err
+}
+
+// Forces a cache reload on next request
+func (c *ClusterClient) forceReload() {
+	atomic.StoreUint32(&c._reload, 1)
+}
+
+// Find the next untried address
+func (c *ClusterClient) findNextAddr(tried map[string]struct{}) string {
+	for addr := range c.addrs {
+		if _, ok := tried[addr]; !ok {
+			return addr
+		}
+	}
+	return ""
+}
+
+// Fetch slot information
+func (c *ClusterClient) fetchClusterSlots(addr string) ([]ClusterSlotInfo, error) {
+	client := NewClient(c.opt.ClientOptions(addr))
+	defer client.Close()
+
+	return client.ClusterSlots().Result()
+}
+
+// Update slot information, populate slots
+func (c *ClusterClient) update(infos []ClusterSlotInfo) {
+	for _, info := range infos {
+		for i := info.Start; i <= info.End; i++ {
+			c.slots[i] = info.Addrs
+		}
+
+		for _, addr := range info.Addrs {
+			c.addrs[addr] = struct{}{}
+		}
+	}
+}
+
+//------------------------------------------------------------------------------
+
+var errNoAddrs = errors.New("redis: no addresses")
+
+type ClusterOptions struct {
+	// A seed-list of host:port addresses of known cluster nodes
+	Addrs []string
+
+	// An optional password
+	Password string
+
+	// The maximum number of MOVED/ASK redirects to follow, before
+	// giving up. Default: 16
+	MaxRedirects int
+
+	// The maximum number of open connections to cluster nodes. Default: 10.
+	// WARNING: Each connection maintains its own connection pool. The maximum
+	// possible number of socket connections is therefore MaxConns x PoolSize
+	MaxConns int
```

Comment:
I understand that you copied this setting from the Ruby client, but do you really think that it is useful? Just wondering.